### DISCUSSION TITLE
Warn: Unused CSS selector

### DISCUSSION THREAD

**Discussion by longrunningprocess** [saadeghi/daisyui#1490] [Q&A] [answered]
Here's my component:

```
<script>
	export let danger = false
	export {clazz as class}

	let clazz = ''
</script>

<!-- https://daisyui.com/components/button -->
<button on:click class:danger class='btn btn-primary { clazz }'>
	<slot />
</button>

<style lang=postcss>
.danger { @apply
	btn-error;
}
</style>
```

When I run `svelte-check`, I get the following:

```
Warn: Unused CSS selector ".btn-outline.danger .badge-outline" (svelte)

<style lang=postcss>
.danger { @apply


Warn: Unused CSS selector ".drawer-toggle:focus-visible ~ .drawer-content .drawer-button.danger" (svelte)

<style lang=postcss>
.danger { @apply
```

If I don't use the `lang=postcss`, I get an "unrecognized" error about the `@apply`.

I was just wondering whether anyone has configured their sveltekit (or svelte-check) in a way that these don't show up.

**Comment by saadeghi** [comment, marked as answer]
Hello again.  
We talked about this before: https://github.com/saadeghi/daisyui/discussions/656#discussioncomment-2455727

>Let's see what's the issue here:
daisyUI has this selector: .label a:hover
and if we don't have a a inside .label it's up to the CSS processor that remove this line or keep it.
Packages like purgecss or uncss will remove this line if .label has no a in it but Tailwind CSS's JIT doesn't remove it.
I think it's because of compatibility reasons. If we have nested components, .label is one file and a is in another file, we wouldn't want that style to be removed just because the CSS preprocessor didn't find the exact hierarchy of styles in a file.

>So, the CSS preprocessor is keeping .label a:hover style. then you use @apply label as a scoped style inside your Svelte component. what happens here is, Tailwind gives .label a:hover style to your component.
Your component checks the style and compares with its own markup. it finds label but can't find <a> inside it, so it warns you that you have unused scoped styles.

This is a similar situation. You're creating abstraction over abstraction.  
Here, daisyUI has `.btn-outline.btn-error` CSS rule. Which is for a button that is both `outline` and the color is `error` color.  
Then you create this scoped CSS in your component
```css
.danger {
  @apply btn-error;
}
```
Which creates `.btn-outline.danger` CSS rule scoped inside your component. Now your component is saying it needs this CSS rule but there's no HTML to use it so it will return a warning.

Generally speaking, scoping CSS + utility first CSS is a bad combination and I wouldn't recommend them together. 
- Utility first libraries like Tailwind CSS are about reusing countless small CSS rules over and over which results smaller CSS file and easier customization.
- Scoping CSS in the component level is about NOT reusing CSS. It's about encapsulating CSS only for 1 component so it wouldn't be reusable.

Please don't use scoping CSS + utility first CSS together. It makes your job harder and it will result a lot of unnecessary duplicate CSS in the production.

**Reply by longrunningprocess** [reply]
yeah, thank you for bringing that to my attention (again 😬 ) and for your patience!
